File: test/ModularPipelines.UnitTests/Plugins/PluginVersionMismatchExceptionTests.cs
using ModularPipelines.Exceptions;

namespace ModularPipelines.UnitTests.Plugins;

public class PluginVersionMismatchExceptionTests
{
    public async Task Message_ContainsPluginName()
    {
        var ex = new PluginVersionMismatchException(
            "ModularPipelines.CustomTool",
            4,
            new Version(5, 2, 0));

        await Assert.That(ex.Message).Contains("ModularPipelines.CustomTool");
    }

    public async Task Message_ContainsRequiredVersion()
    {
        var ex = new PluginVersionMismatchException(
            "TestPlugin",
            4,
            new Version(5, 2, 0));

        await Assert.That(ex.Message).Contains("4.x");
    }

    public async Task Message_ContainsActualVersion()
    {
        var ex = new PluginVersionMismatchException(
            "TestPlugin",
            4,
            new Version(5, 2, 0));

        await Assert.That(ex.Message).Contains("5.2.0");
    }

    public async Task Message_HandlesNullVersion()
    {
        var ex = new PluginVersionMismatchException(
            "TestPlugin",
            4,
            null);

        await Assert.That(ex.Message).Contains("unknown");
    }

    public async Task Properties_AreSetCorrectly()
    {
        var version = new Version(5, 2, 0);
        var ex = new PluginVersionMismatchException("TestPlugin", 4, version);

        using (Assert.Multiple())
        {
            await Assert.That(ex.PluginName).IsEqualTo("TestPlugin");
            await Assert.That(ex.RequiredMajorVersion).IsEqualTo(4);
            await Assert.That(ex.ActualCoreVersion).IsEqualTo(version);
        }
    }
}

File: test/ModularPipelines.UnitTests.FSharp/Plugins/PluginVersionMismatchExceptionTests.fs
namespace ModularPipelines.UnitTests.FSharp.Plugins

open ModularPipelines.UnitTests.Plugins
open ModularPipelines.UnitTests.FSharp
open TUnit.Core

type PluginVersionMismatchExceptionTests() =
    inherit ModularPipelines.UnitTests.Plugins.PluginVersionMismatchExceptionTests()

    [<Test>]
    member this.Message_ContainsPluginName() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Plugins.PluginVersionMismatchExceptionTests> "Message_ContainsPluginName" None

    [<Test>]
    member this.Message_ContainsRequiredVersion() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Plugins.PluginVersionMismatchExceptionTests> "Message_ContainsRequiredVersion" None

    [<Test>]
    member this.Message_ContainsActualVersion() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Plugins.PluginVersionMismatchExceptionTests> "Message_ContainsActualVersion" None

    [<Test>]
    member this.Message_HandlesNullVersion() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Plugins.PluginVersionMismatchExceptionTests> "Message_HandlesNullVersion" None

    [<Test>]
    member this.Properties_AreSetCorrectly() =
        CSharpTestWrapper.invokeTest this typeof<ModularPipelines.UnitTests.Plugins.PluginVersionMismatchExceptionTests> "Properties_AreSetCorrectly" None
